### DISCUSSION TITLE
meaning of ":="

### DISCUSSION THREAD

**Discussion by wdduncan** [EvgSkv/logica#215] [General]
In the file `lib/closure.l `, I see the line:
```
TransitiveClosure := TC4();
```
Does `:=` mean assignment?

**Comment by EvgSkv** [comment]
`:=` operator indeed acts like an assignment, but it has a more specific meaning. It initiates so called `functor call`. We do need better documentation on this. For now I've added a section in the tutorial: [Functors](https://colab.research.google.com/github/EvgSkv/logica/blob/main/tutorial/Logica_tutorial.ipynb#scrollTo=80nzV8_mbXFA).

Functor calls are formally second order functions. They are also similar to SQL's macros.
You can take any predicate `P` and make its version, where arbitrary predicates used in definition of `P` are replaced with you some other predicates. This lets re-using complex logic, which isn't easy to re-use with regular predicate arguments.

Let me know if you have further questions.

**Reply by wdduncan** [reply]
Hi @EvgSkv Thanks for the reply. It helps somewhat. I'm trying to understand your code in `lib/closure.l `

First, let's start here:
```
TC0(a, b) distinct :- R(a, b);

@Ground(TC1);
TC1(a, b) distinct :- TC0(a, b);
TC1(a, c) distinct :- TC0(a, b), TC0(b, c);
```
`@Ground` associates a predicate with a table (as described [here](https://colab.research.google.com/github/EvgSkv/logica/blob/main/tutorial/Logica_tutorial.ipynb#scrollTo=ZvDxW-sSZE6Q)). But, I'm wondering why you couldn't/didn't do this:
```
TC1(a, b) distinct :- R(a, b);
TC1(a, c) distinct :- R(a, b), R(b, c);
```
In other words, do you need this line?: `TC0(a, b) distinct :- R(a, b);`  

In the remaining lines, you use functors:
```
TC2 := TC1(TC0: TC1);
TC3 := TC2(TC0: TC2);
TC4 := TC3(TC0: TC3);
```
Does `TC2 := TC1(TC0: TC1);` mean the logic for satisfying the goal `TC2` is created by substituting the `TC0` predicate in `TC1` with `TC1`? I'm finding this confusing.

Also, does `TransitiveClosure := TC4();` effectively make `TransitiveClosure` the same as `TC4`? That is, `TransitiveClosure` is satisfied whenever `TC4` is satisfied, and vice versa.

**Reply by EvgSkv** [reply]
@wdduncan thanks for your questions! I'm happy to go over them, but for clarity I wanted to mention first that this example was written before recursion was implenented in Logica. Now it's left for reference when recursion is discussed. For the transitive closure specifically nowadays one should just use the build-in recursion.
[Here is an example](https://colab.research.google.com/drive/1s97Uihqj3LuE_6wx915AELd60rRZTrhM#scrollTo=oT338S6Fsvfr) in the colab of using TransitiveClosure functor that your questions refer to to implenent Ancestor. And there is Ansestor2 predicate implemented via recursion, which is the encouraged approach.

Now to your questions.

> @Ground associates a predicate with a table (as described [here](https://colab.research.google.com/github/EvgSkv/logica/blob/main/tutorial/Logica_tutorial.ipynb#scrollTo=ZvDxW-sSZE6Q)).

Correct!

> But, I'm wondering why you couldn't/didn't do this:
> ...

I guess it could have been done like that. `TC0` was used to emphasise that R is the first step of the recursion.

>Does TC2 := TC1(TC0: TC1); mean the logic for satisfying the goal TC2 is created by substituting the TC0 predicate in TC1 with TC1?

Yes, that's exactly right! TC1 is true for pairs for which TC0 is true and for pairs (a,c) for which exist b such that TC0(a,b), TC0(b,c).
And TC2 is obtained by replacing TC0 with TC1, thus we get definition:
_TC2 is true for pairs for which TC1 is true and for pairs (a,c) for which exist b such that TC1(a,b), TC1(b,c)._

This is basically a way to iterate the predicate.

>Also, does TransitiveClosure := TC4(); effectively make TransitiveClosure the same as TC4? That is, TransitiveClosure is satisfied whenever TC4 is satisfied, and vice versa.

Correct! This is just renaming TC4 into a nicer named TransitiveClosure.

I'm happy to discuss this in further details, but again, in practice you can just use recursion and think of it as magic :-)
